### ISSUE TITLE
metadata returned by plugin is inconsistent happens

### ISSUE THREAD

**Issue by sapphi-red** [evanw/esbuild#2177]
Maybe Related: #2176, https://github.com/vitejs/vite/issues/7683, https://github.com/vitejs/vite/issues/7685

## Reproduction
1. Open https://stackblitz.com/edit/node-d2wvrw?file=bundle.js
2. Download it
3. Run `npm i` and `node bundle.js`

On windows, `Detected inconsistent metadata for the path "src/c.js" when it was imported here:` happens.
But on stackblitz, it does not happen.

## Full error log
```text
X [ERROR] Detected inconsistent metadata for the path "src/c.js" when it was imported here:

    src/b.js:1:7:
      1 │ import './c.js';
        ╵        ~~~~~~~~

  The original metadata for that path comes from when it was imported here:

    src/a.js:1:7:
      1 │ import '@/c.js';
        ╵        ~~~~~~~~

  The difference in metadata is displayed below:

   {
  -  "pluginName": "alias",
  -  "type": null,
  +  "pluginName": null,
  +  "type": "module",
   }

  This is a bug in the "alias" plugin. Plugins provide metadata for a given path in an "onResolve"
  callback. All metadata provided for the same path must be consistent to ensure deterministic
  builds. Due to parallelism, one set of provided metadata will be randomly chosen for a given path,       
  so providing inconsistent metadata for the same path can cause non-determinism.

C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1603
  let error = new Error(`${text}${summary}`);
              ^

Error: Build failed with 1 error:
src/b.js:1:7: ERROR: Detected inconsistent metadata for the path "src/c.js" when it was imported here:     
    at failureErrorWithLog (C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1603:15) 
    at C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1249:28
    at runOnEndCallbacks (C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1034:63)   
    at buildResponseToResult (C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1247:7)    at C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:1356:14
    at C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:666:9
    at handleIncomingPacket (C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:763:9)  
    at Socket.readFromStdout (C:\Users\green\Downloads\node-d2wvrw\node_modules\esbuild\lib\main.js:632:7) 
    at Socket.emit (node:events:526:28)
    at addChunk (node:internal/streams/readable:315:12) {
  errors: [
    {
      detail: undefined,
      location: {
        column: 7,
        file: 'src/b.js',
        length: 8,
        line: 1,
        lineText: "import './c.js';",
        namespace: '',
        suggestion: ''
      },
      notes: [
        {
          location: {
            column: 7,
            file: 'src/a.js',
            length: 8,
            line: 1,
            lineText: "import '@/c.js';",
            namespace: '',
            suggestion: ''
          },
          text: 'The original metadata for that path comes from when it was imported here:'
        },
        {
          location: null,
          text: 'The difference in metadata is displayed below:'
        },
        { location: null, text: '' },
        { location: null, text: ' {' },
        { location: null, text: '-  "pluginName": "alias",' },
        { location: null, text: '-  "type": null,' },
        { location: null, text: '+  "pluginName": null,' },
        { location: null, text: '+  "type": "module",' },
        { location: null, text: ' }' },
        { location: null, text: '' },
        {
          location: null,
          text: 'This is a bug in the "alias" plugin. Plugins provide metadata for a given path in an "onResolve" callback. All metadata provided for the same path must be consistent to ensure deterministic builds. Due to parallelism, one set of provided metadata will be randomly chosen for a given path, so providing inconsistent metadata for the same path can cause non-determinism.'
        }
      ],
      pluginName: '',
      text: 'Detected inconsistent metadata for the path "src/c.js" when it was imported here:'
    }
  ],
  warnings: []
}
```

**Comment by sapphi-red** [comment]
This also happens with https://stackblitz.com/edit/node-dmisef?file=bundle.js.

plugin of first reproduction
```js
const aliasPlugin = {
  name: 'alias',
  setup({ onResolve }) {
    onResolve({ filter: /^@/ }, ({ path: p, importer }) => {
      return {
        path: path.resolve(path.dirname(importer), p.replace(/^@/, '.')),
      };
    });
  },
};
```

plugin of second reproduction
```js
const aliasPlugin = {
  name: 'alias',
  setup({ onResolve, resolve }) {
    onResolve({ filter: /^@/ }, async ({ path: p, importer }) => {
      return await resolve(p.replace(/^@/, '.'), {
        resolveDir: path.dirname(importer),
      });
    });
  },
};
```

**Comment by acidjazz** [comment]
similar issues 

https://github.com/vitejs/vite/issues/7685

https://github.com/vitejs/vite/issues/7683

**Comment by alex-jungle-cat** [comment]
can reproduce on Linux as well

**Comment by sapphi-red** [comment]
I confirmed this happens with linux by using docker.
Maybe it cannot be reprocuced only with stackblitz.

**Comment by sapphi-red** [comment]
The diff of `"type": "module",` was happening because `finalizeResolve` is not called when it is resolved by a plugin.
https://github.com/evanw/esbuild/blob/757d7d999b7ad1a65636c1fe34f21ec18a445b01/internal/bundler/bundler.go#L788-L794
https://github.com/evanw/esbuild/blob/757d7d999b7ad1a65636c1fe34f21ec18a445b01/internal/resolver/resolver.go#L654-L782

**Comment by evanw** [comment]
Yes, I can confirm locally as well. Thanks for the report and for the suggested fix. This clearly should have been introduced as a warning instead of an error though. My apologies.

I believe what esbuild is flagging here may be representative of a real issue that was previously undiscovered. The discrepancy between the module types could hypothetically lead to a divergence in build behavior in this case depending on which set of metadata is chosen first. I will revert this for now and then work on getting this fixed.

**Comment by evanw** [comment]
The revert has been published as 0.14.36, so that version should work. Sorry about the breakage.

A bit more background: As a performance shortcut, esbuild's internals pass metadata about the resolved file (e.g. `module` and `sideEffects` from `package.json`) from the path resolution phase directly to the parsing phase instead of re-discovering it when parsing a file. That way it doesn't need to be looked up again. There isn't a concern about non-determinism when esbuild does this because path resolution always runs the same code and so is always consistent.

However, when the plugin API was added that design wasn't changed, and is error-prone as a result because plugins can inject arbitrary code that doesn't follow the original rules. I could make the suggested change above to call `finalizeResolve` but that would overwrite these metadata values from `onResolve` plugins and make it not possible to set metadata via plugins anymore. Instead I could consider making a breaking change to the plugin API to move this metadata into either `onLoad` or into another plugin callback to decorate a file with metadata. This wouldn't be error-prone because it would only happen once per file, but it would be slower because it would involve more communication overhead and delays in the pipeline.
